src/foam/nanos/ruler/RuleDebugger.js

```diff
@@ -0,0 +1,151 @@
+    },
+    {
+      class: 'Int',
+      name: 'priority',
+      value: 2147483647,
+      documentation: 'Ensure this rule always runs first'
+    },
+    {
+      class: 'Enum',
+      of: 'foam.nanos.dao.Operation',
+      name: 'operation',
+      value: 'CREATE_OR_UPDATE'
+    }
+  ],
+
+  methods: [
+    {
+      name: 'applyAction',
+      javaCode: `
+        x = getX();
+        DAO rulerDAO = (DAO) x.get("ruleDAO");
+        DAO ruleGroupDAO = (DAO) x.get("ruleGroupDAO");
+        AuthService auth = (AuthService) x.get("auth");
+
+        List rules = ((ArraySink) rulerDAO.where(
+          AND(
+            EQ(Rule.DAO_KEY, getDaoKey()),
+            NEQ(Rule.ID, getId()),
+            EQ(Rule.OPERATION, getOperation()),
+            EQ(Rule.AFTER, getAfter())
+          )).select(new ArraySink())).getArray();
+
+        StringBuilder sb = new StringBuilder();
+        sb.append("Performing Rule Debug on DAO: ");
+        sb.append(getDaoKey());
+        sb.append(" from rule: ");
+        sb.append(getId());
+        sb.append("\\n");
+        Logger logger = (Logger) x.get("logger");
+
+        if ( rules == null || rules.size() == 0 ) {
+          sb.append("Error No Rules Found in DAOKey ");
+          logger.error(sb.toString());
+          throw new RuntimeException(sb.toString());
+        }
+        sb.append("\\nFound ");
+        sb.append(rules.size());
+        sb.append(" rules");
+        for ( Object o : rules ) {
+          Rule r = (Rule) o;
+          RuleGroup group = (RuleGroup)ruleGroupDAO.find(r.getRuleGroup());
+          if ( group == null ) {
+            sb.append(" unable to find ruleGroup ");
+            continue;
+          }
+          sb.append(" group: ");
+          sb.append(group.getId());
+          if ( ! r.getEnabled() ) {
+            sb.append(" Disabled: ");
+            sb.append(r.getName());
+            sb.append(" ");
+            sb.append(r.getId());
+            continue;
+          }
+          sb.append(" Enabled: ");
+          sb.append(r.getName());
+          sb.append(" ");
+          sb.append(r.getId());
+          if ( ! group.getEnabled() ) {
+            sb.append(" Group is DISABLED");
+            continue;
+          }
+
+          // --- Permission check ---
+
+          User user = r.getUser(x, obj);
+          sb.append("\\nRULE PERMISSION GRANTED: ");
+          if ( user != null ) {
+            sb.append(auth.checkUser(x, user, "rule.read." + r.getId()));
+            sb.append("\\nRULEGROUP PERMISSION GRANTED: ");
+            sb.append(auth.checkUser(x, user, "rulegroup.read." + group.getId()));
+          }
+          else {
+            sb.append("NOT REQUIRED");
+          }
+
+          // --- Predicate check ---
+
+          sb.append("\\nPREDICATE CHECK: ");
+          sb.append(r.getPredicate().f(obj));
+          sb.append("\\nGROUP PREDICATE CHECK: ");
+          sb.append(group.getPredicate().f(obj));
+        }
+
+      logger.debug(sb.toString());
```

Comment:
Also ideally would be nice to take into consideration "stop" functionality but not possible to do it here. RuleEngine has much more info about rules and whats being executed, would be good to create DebugRuleEngine or something like that and use that for debugging purposes, idea for todo? @jlhughes @kgrgreer @lchanmann